---
title: "Class 14"
author: "Jason Baer"
date: "11/13/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r eval = FALSE}
intall.packages("BiocManager")
BiocManager::install("DESeq2")
library(BiocManager)
library(DESeq2)
```


```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
counts

```

```{r}
metadata <- read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
metadata

```

Q1: How many genes are in the dataset?

```{r}
nrow(counts)
```

How many experiments?

```{r}
ncol(counts) - 1
```

Lets make sure metadata id col matches the colnames of counts

```{r}
all(colnames(counts)[-1] == metadata$id )
```


Q2: How many control cell lines do we have?

```{r}
metadata$dex == "control"
```

First we want to access the columns of countData that are control and then find their mean.

```{r}
control.id <- metadata[metadata$dex == "control", ]$id
control.id
```


```{r}
control.mean <- rowSums(counts[, control.id])/length(control.id)
names(control.mean) <- counts$ensgene
head(control.mean)
```


Then we can do the same for treated samples

```{r}
treated.id <- metadata[metadata$dex == "treated", ]$id
treated.id
```


```{r}
treated.mean <- rowSums(counts[ ,treated.id]) / length(treated.id)
names(treated.mean) <- counts$ensgene
head(treated.mean)
```

Combine meancount data into a dataframe

```{r}
meancounts <- data.frame(control.mean, treated.mean)
meancounts
```

Plot control vs treated

```{r}
plot(meancounts$control.mean, meancounts$treated.mean, xlab = "Control Mean", ylab = "Treated Mean")
```

Plot both of these on a log scale

```{r}
plot(meancounts$control.mean, meancounts$treated.mean, xlab = "Control Mean", ylab = "Treated Mean", log = "xy")
```

Fold change. negative numbers mean fold change is going down, positive mean it is going up. -inf means negative infinity (taking log of zero), so genes that have zero expression. 

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

Lets filter out genes with no expression and NaN

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
nrow(mycounts)
```


Test for finding zero entries

```{r}
x <- data.frame(happy = c(5,6,0,0), sad = c(0,5,5,0))
x == 0
which(x == 0)

which(x == 0, arr.ind = TRUE)
```

A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Letâ€™s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
length(which(up.ind, arr.ind = TRUE)) #can also use sum(up.ind)
length(which(down.ind, arr.ind = TRUE))
```

```{r}
anno <- read.csv("annotables_grch38.csv")
head(anno)
```

We will use Bioconductor's annotation packages to help with mapping various ID schemes to each other. Here we load the AnnotationDBI package and the annotation package org.Hs.eg.db. Lets see whats available in org.Hs


```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
columns(org.Hs.eg.db)
```

We will add a new column in mycounts called symbol. 

```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```

```{r}
head(mycounts)
```

Run the mapIds() function two more times to add the Entrez ID and UniProt accession as new columns called mycountsentrez and mycountsuniprot

```{r}
mycounts$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")
```

```{r}
mycounts$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")

head(mycounts)
```


```{r}
library(DESeq2)
citation("DESeq2")
```


```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

running DESeq main package

```{r}
dds <- DESeq(dds)
```

call results

```{r}
res <- results(dds)
res
```

#volcano plot

```{r}
plot(res$log2FoldChange, res$padj)
```





```{r}
plot(res$log2FoldChange, -log(res$padj))

```


To color the points we will setup a custom color vector indicating transcripts with large fold change and significant differences between conditions

```{r}
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```



